### PR TITLE
Parallelize metadata syncing on node activate

### PR DESCRIPTION
**Pull request by onderkalaci**
It is often useful to be able to sync the metadata in parallel
across nodes.

Note that this commit does not parallelize all pieces of node
activation or metadata syncing. Instead, it tries to parallelize
potenially large parts of metadata, which is the objects and
distributed tables (in general Citus tables).

In the future, it would be nice to sync the reference tables
in parallel across nodes.

TODO:
- [ ] Add some more explicit tests (although we have some through `citus_finalize_upgrade_to_citus11`)

Create ~720 distributed tables / ~23450 shards
```SQL
-- declaratively partitioned table
CREATE TABLE github_events_looooooooooooooong_name (
  event_id bigint,
  event_type text,
  event_public boolean,
  repo_id bigint,
  payload jsonb,
  repo jsonb,
  actor jsonb,
  org jsonb,
  created_at timestamp
) PARTITION BY RANGE (created_at);

SELECT create_time_partitions(
  table_name         := 'github_events_looooooooooooooong_name',
  partition_interval := '1 day',
  end_at             := now() + '24 months'
);

CREATE INDEX ON github_events_looooooooooooooong_name USING btree (event_id, event_type, event_public, repo_id);
SELECT create_distributed_table('github_events_looooooooooooooong_name', 'repo_id');

SET client_min_messages TO ERROR;

```

across 1 node: almost same as expected
```SQL

SELECT start_metadata_sync_to_primary_nodes();
Time: 15664.418 ms (00:15.664)

select start_metadata_sync_to_node(nodename,nodeport) from pg_dist_node;
Time: 14284.069 ms (00:14.284)
```

across 7 nodes: ~3.5x improvement
```SQL

SELECT start_metadata_sync_to_primary_nodes();
┌──────────────────────────────────────┐
│ start_metadata_sync_to_primary_nodes │
├──────────────────────────────────────┤
│ t                                    │
└──────────────────────────────────────┘
(1 row)

Time: 25711.192 ms (00:25.711)

-- across 7 nodes
select start_metadata_sync_to_node(nodename,nodeport) from pg_dist_node;
Time: 82126.075 ms (01:22.126)
```

### CODE REVIEW

File: src/backend/distributed/transaction/worker_transaction.c [hunk 1]
```diff
@@ -490,27 +490,72 @@ SendCommandListToWorkerOutsideTransaction(const char *nodeName, int32 nodePort,
  * coordinated transaction. Any failures aborts the coordinated transaction.
  */
 void
-SendMetadataCommandListToWorkerInCoordinatedTransaction(const char *nodeName,
-														int32 nodePort,
-														const char *nodeUser,
-														List *commandList)
+SendMetadataCommandListToWorkerListInCoordinatedTransaction(List *workerNodeList,
```
Comment (onderkalaci) [inline src/backend/distributed/transaction/worker_transaction.c:493]: @marcocitus do you see any drawbacks of this function? I'm mostly asking regarding failure handling. All seems good to me, but I'd appreciate if you can also have a quick look

File: src/test/regress/expected/failure_mx_metadata_sync.out [hunk 2]
```diff
@@ -51,70 +52,58 @@ SELECT citus.mitmproxy('conn.onQuery(query="^UPDATE pg_dist_local_group SET grou
...
 ---------------------------------------------------------------------
 
 (1 row)
 
 SELECT start_metadata_sync_to_node('localhost', :worker_2_proxy_port);
-ERROR:  server closed the connection unexpectedly
-	This probably means the server terminated abnormally
-	before or while processing the request.
-CONTEXT:  while executing command on localhost:xxxxx
+ERROR:  failure on connection marked as essential: localhost:xxxxx
```
Comment (onderkalaci) [inline src/test/regress/expected/failure_mx_metadata_sync.out:106]: I think we should consider improving these error messages in general for essential connection: https://github.com/citusdata/citus/issues/5515

File: src/backend/distributed/metadata/node_metadata.c [hunk 3]
```diff
@@ -1118,15 +1136,14 @@ PrimaryNodeForGroup(int32 groupId, bool *groupContainsNodes)
...
- * given node.
+ * ActivateNodeList iterates over the nodeList and activates the nodes.
+ * Some part of the node activation is done parallel across the nodes,
+ * such as syncing the metadata. However, reference table replication is
+ * done one by one across nodes.
  */
-int
-ActivateNode(char *nodeName, int nodePort)
+void
+ActivateNodeList(List *nodeList)
```
Comment (onderkalaci) [inline src/backend/distributed/metadata/node_metadata.c:1145]: in the future, we can invest more into this function and provide activation of multiple nodes in parallel (including reference table replication).
For now, we only use multiple node activation for the new UDF

File: src/backend/distributed/metadata/metadata_sync.c [hunk 4]
```diff
@@ -195,6 +196,33 @@ start_metadata_sync_to_node(PG_FUNCTION_ARGS)
 }
 
 
+/*
+ * start_metadata_sync_to_primary_nodes function sets hasmetadata column of
```
Comment (onderkalaci) [inline src/backend/distributed/metadata/metadata_sync.c:200]: We should perhaps use a better name like: `resync_metadata_to_primary_nodes` or such?

Comment (marcocitus) [reply src/backend/distributed/metadata/metadata_sync.c:200]: "primary" kind of suggest that you could sync to a different type of node as well. Maybe start_metadata_sync / start_metadata_sync_to_all_nodes?

We'll implicitly replicate to standby nodes.

File: src/backend/distributed/transaction/worker_transaction.c [hunk 5]
```diff
@@ -490,27 +490,72 @@ SendCommandListToWorkerOutsideTransaction(const char *nodeName, int32 nodePort,
...
+						 linitial(commandList) : StringJoin(commandList, ';');
+
+	/* send commands in parallel */
+	bool failOnError = true;
+	MultiConnection *connection = NULL;
+	foreach_ptr(connection, connectionList)
 	{
-		ExecuteCriticalRemoteCommand(workerConnection, commandString);
+		int querySent = SendRemoteCommand(connection, stringToSend);
+		if (failOnError && querySent == 0)
```
Comment (marcocitus) [inline src/backend/distributed/transaction/worker_transaction.c:550]: failOnError seems always true

File: src/backend/distributed/transaction/worker_transaction.c [hunk 6]
```diff
@@ -490,27 +490,72 @@ SendCommandListToWorkerOutsideTransaction(const char *nodeName, int32 nodePort,
...
+	foreach_ptr(workerNode, workerNodeList)
+	{
+		const char *nodeName = workerNode->workerName;
+		int nodePort = workerNode->workerPort;
+		int connectionFlags = REQUIRE_METADATA_CONNECTION;
+
+		MultiConnection *connection = StartNodeUserDatabaseConnection(connectionFlags,
 																	  nodeName, nodePort,
-																	  nodeUser, NULL);
+																	  CurrentUserName(),
```
Comment (marcocitus) [inline src/backend/distributed/transaction/worker_transaction.c:516]: minor, can just use StartNodeConnection I think

File: src/backend/distributed/metadata/metadata_sync.c [hunk 7]
```diff
@@ -195,6 +196,33 @@ start_metadata_sync_to_node(PG_FUNCTION_ARGS)
...
+ */
+Datum
+start_metadata_sync_to_primary_nodes(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+
+	EnsureSuperUser();
+	EnsureCoordinator();
+
+	List *workerNodes = ActivePrimaryNonCoordinatorNodeList(ShareLock);
```
Comment (marcocitus) [inline src/backend/distributed/metadata/metadata_sync.c:212]: I think we generally need RowShareLock 

(there are some cases where we take ShareLock, iirc this causes some issues)

File: src/test/regress/expected/failure_add_disable_node.out [hunk 8]
```diff
@@ -106,10 +106,11 @@ SELECT citus.mitmproxy('conn.onQuery(query="CREATE SCHEMA").kill()');
 (1 row)
 
 SELECT master_activate_node('localhost', :worker_2_proxy_port);
-ERROR:  server closed the connection unexpectedly
+WARNING:  server closed the connection unexpectedly
 	This probably means the server terminated abnormally
 	before or while processing the request.
 CONTEXT:  while executing command on localhost:xxxxx
+ERROR:  failure on connection marked as essential: localhost:xxxxx
```
Comment (marcocitus) [inline src/test/regress/expected/failure_add_disable_node.out:113]: small downside: this error is a bit worse

**Review by marcocitus — APPROVED** [review]

**Pull request merged**